Class Mentoring Rework

zumbak04 opened the pull request:
## Changelog:
- You can ask characters to be class mentors of other characters like you children. Some mentors would like not to take students because of their traits or because of hostility.

## Developer changelog:
- The correspondence between a class mentor and student takes more time.
- Added a fallback/cleanup event in case of death of a class mentor or student during their correspondence.
- Rewrote the localization of the mentoring events (the correspondence, decision description etc.).

## Script changelog (ignore if you can't read CK2 script):
- Added `ai_chance` in `WCCLS.1500` to make sure the AI prefers to pick a class, not cancel the event.
- Removed `opinion_refused_to_be_mentor` from `WCCLS.1500` event.
- Made changes in the Hidden Class mechanics. Added `this_see_prev_class_trait_trigger` trigger.
- Renamed `opinion_rejected_request_for_mentoring` into `opinion_refused_to_be_mentor` because the name was too long.

## How to Test
Try the new features:
- Try to assign mentors to your children.
- Try to select mentors for yourself.
- Make sure the event, where an Educator trains your children when they reach 12 years old, works fine.  You may find 11 years old, assign an Educator with high class level to her/him, wait for the event.
- Make sure the Hidden Class mechanics work. You may test it with mages in the Kaldorei Empire. They start with banned Arcane magic, so mages shouldn't be visible. If you're not mage yourself, you shoudn't see their mana, class trait. You shouldn't be able to `Ask to be Mentor` as well. Though you should be able to `Ask to be Mentor` if your religion is... Sunwell... for example. `religion cult_of_sunwell` in the console to change your religion to Sunwell.

After the mentor has agreed to train you (or your children), make sure `ask_to_be_mentor_ongoing_flag` flag is cleared on you (or your children).

MemeWizard commented:
So, you can ask character to become mentor of your landed vassal even if you have banned %magictype% and will not be able to see new class trait after completing education.

<details>
<summary>Click to expand</summary>

![image](https://user-images.githubusercontent.com/46576860/69189224-01d00580-0b2f-11ea-846d-4563ff0b229a.png)
![image](https://user-images.githubusercontent.com/46576860/69189538-963a6800-0b2f-11ea-95f6-c657709cb56f.png)
![image](https://user-images.githubusercontent.com/46576860/69190024-9d15aa80-0b30-11ea-9aa6-91edc0ee6c19.png)

</details>

Double scope
<details>
<summary>Click to expand</summary>

![image](https://user-images.githubusercontent.com/46576860/69190160-da7a3800-0b30-11ea-9414-287701263b89.png)

</details>

Also, notification about characters completing their education would be nice.

zumbak04 commented:
> So, you can ask character to become mentor of your landed vassal even if you have banned %magictype% and will not be able to see new class trait after completing education.

Fixed in https://github.com/Warcraft-GoA-Development-Team/Warcraft-Guardians-of-Azeroth/pull/948/commits/3a2cb8607cf321faa8911603bad6b4a0bd9d5e2c ! Needs more testing.

> Double scope

Fixed in https://github.com/Warcraft-GoA-Development-Team/Warcraft-Guardians-of-Azeroth/pull/948/commits/3a2cb8607cf321faa8911603bad6b4a0bd9d5e2c !

zumbak04 commented:
> Also, notification about characters completing their education would be nice.

Implemented!

MemeWizard commented on decisions/wc_class_decisions.txt:
```diff
@@ -1155,38 +1155,223 @@ decisions = {
+				value = -100
+				localisation_key = ROOT_dislikes_FROM_tooltip
+				opinion = { who = FROM value == -100 }
+			}
```
How about make it 10 per 10 instead of 20 per 20?
<details>
<summary>Click to expand</summary>

![image](https://user-images.githubusercontent.com/46576860/70449993-0182b480-1ab4-11ea-8916-4449be255a0d.png)
![image](https://user-images.githubusercontent.com/46576860/70450025-0e070d00-1ab4-11ea-8912-4410c1e801e0.png)


</details>

MemeWizard replied:
Added!
@zumbak04, check them, please

zumbak04 commented:
> It's weird that my character will think bad about mentor, because mentor refused educating your character, but it's my char who refused to do it.

@MemeWizard have an idea of renaming it... something like `Negotiations Failed`.

MemeWizard commented:
It still will be weird, since it's just my character who changed it's mind
Also, there was no negotiations, so it will be even weirder

zumbak04 commented:
> It's weird that my character will think bad about mentor, because mentor refused educating your character, but it's my char who refused to do it.

Mmmkay, removed the opinion.

zumbak04 commented on localisation/00_zuwc_united_localisation.csv:
```diff
@@ -1943,7 +1926,6 @@ underground;Underground;Souterrains;;;;;;;;;;;;x
 evil_true_religion_trigger_tooltip;Serves Legion, Old Gods or Lich King\n;Sert la L�gion, les Dieux tr�s anciens ou le roi-liche\n;;;;;;;;;;;;x
 has_unpleasant_trait_trigger_tooltip;Has unpleasant trait (deceitful, shy, paranoid, cruel, stubborn, haughty, brooding, impaler, greedy, wroth, envious etc.)\n;;;;;;;;;;;;;x
 is_my_educator_tooltip;Is my educator\n;;;;;;;;;;;;;x
```
Changed in https://github.com/Warcraft-GoA-Development-Team/Warcraft-Guardians-of-Azeroth/pull/948/commits/51429a05aefaeb5ee11451a7bb88640588a7e62f.

zumbak04 commented on common/scripted_effects/wc_class_scripted_effects.txt:
```diff
@@ -616,7 +616,7 @@ add_class_trait_effect = {
 			# limit = { true_religion_group = void_group }
 			# add_trait = subclass_dark_shaman
-		# }
+			# }
```
Fixed in https://github.com/Warcraft-GoA-Development-Team/Warcraft-Guardians-of-Azeroth/pull/948/commits/bbdbc3d76791749a6075c7bf5a4288ca7d7f943f.

zumbak04 commented on common/scripted_effects/wc_class_scripted_effects.txt:
```diff
@@ -1922,8 +1922,8 @@ ai_add_random_class_effect = {
 					# modifier = {
 						# factor = 0
 						# united_death_knight_trigger = no
+						# }
```
Fixed in https://github.com/Warcraft-GoA-Development-Team/Warcraft-Guardians-of-Azeroth/pull/948/commits/bbdbc3d76791749a6075c7bf5a4288ca7d7f943f.

zumbak04 commented on interface/domestic_diplomacy.gui:
```diff
@@ -0,0 +1,1521 @@
+					maxHeight = 20
+					fixedsize = yes
+					format = center
+				}
```
@MemeWizard you play on 3.3.0?

MemeWizard replied:
Yep, but I can't reproduce it.
Sometimes it bugged, sometimes not.

zumbak04 replied:
Paradox bug. There must be either Yes, No or Maybe, but there was empty space, so I added at least some text.
Never seen it fine on 3.2.2. Once was on 3.3.0.

MemeWizard replied:
So, it's ok that this thing pop-up in every interaction with third character, even if AI can't refuse?

<details>
<summary>Click to expand</summary>

![image](https://user-images.githubusercontent.com/46576860/73117580-f6e38800-3f58-11ea-9b0d-a4641b8e1537.png)
![image](https://user-images.githubusercontent.com/46576860/73117585-006cf000-3f59-11ea-9be8-c13dadc6a50b.png)
![image](https://user-images.githubusercontent.com/46576860/73117593-124e9300-3f59-11ea-902f-bc42b92430ec.png)

</details>

Also, I think this bug is caused by another interaction with third character.
For example, if you ask character to join plot, its reply will merge with this thing.

<details>
<summary>Click to expand</summary>

![image](https://user-images.githubusercontent.com/46576860/73117671-6efe7d80-3f5a-11ea-8743-6b28cb640b11.png)
![image](https://user-images.githubusercontent.com/46576860/73117709-b5ec7300-3f5a-11ea-9f25-b238bd8f3415.png)
![image](https://user-images.githubusercontent.com/46576860/73117713-be44ae00-3f5a-11ea-97a0-0f38992f93d0.png)

</details>

zumbak04 replied:
> So, it's ok that this thing pop-up in every interaction with third character, even if AI can't refuse?

Don't think so, but I doubt Paradox is going to fix the interface at this point.

MemeWizard replied:
@zumbak04, looks like I fixed that problem in dfaf054, check it too

zumbak04 commented on common/scripted_triggers/wcsp_scripted_triggers.txt:
```diff
@@ -1,39 +1,40 @@
 	is_training_self_trigger = no
 	is_training_mentor_trigger = no
+	
+	# Evil characters can only train characters of their religion
```
As I know, there's no way to hide the decision if there are no characters without hurting performance.

MemeWizard replied:
Maybe instead of "if there are no characters", hide the decision if decision target is evil and character that press the decision don't have the same religion as decision target?

zumbak04 replied:
Changed in https://github.com/Warcraft-GoA-Development-Team/Warcraft-Guardians-of-Azeroth/pull/948/commits/e7041ec06017aaac5e48ae02f36346c7aa0826bb.

zumbak04 commented:
> It's impossible to choose yourself as mentor

Planned in the next PR.

> You can ask character to become mentor of multiple characters if it don't have pupils, but can't ask character if it have one.

Fixed in https://github.com/Warcraft-GoA-Development-Team/Warcraft-Guardians-of-Azeroth/pull/948/commits/49290976a7268cfb4ec2a6ddc9aafc346897e2bc.

MemeWizard commented:
<details>
<summary>Click to expand</summary>

![image](https://user-images.githubusercontent.com/46576860/73117730-f4822d80-3f5a-11ea-85f3-de4baeadf78a.png)

</details>

MemeWizard commented:
> You no longer see **Ask to be Mentor** decisions by clicking evil characters.

What about vice versa?

<details>
<summary>Click to expand</summary>

![image](https://user-images.githubusercontent.com/46576860/74797025-3d6b9e80-52db-11ea-845e-86a8f7116146.png)

</details>

MemeWizard commented:
Suggestion:
Add "master of dagger" for rogues

<details>
<summary>Click to expand</summary>

![image](https://user-images.githubusercontent.com/46576860/74923155-58283b00-53e1-11ea-9cc0-7654c990cda4.png)

</details>

MemeWizard commented on localisation/00_zuwc_united_localisation.csv:
```diff
@@ -2884,6 +2866,38 @@ EVTOPTA_WCUND_9915;Bow to the Death God!;;;;;;;;;;;;;x
+EVTDESC_WCCLS_2103_others;[Root.PrimaryTitle.GetFOA], apologies, but I have no desire be a mentor of [student.GetDynName].\n\nSigned,\n[From.GetTitledName];;;;;;;;;;;;;x
+has_not_students_tooltip;Does NOT have students;;;;;;;;;;;;;x
+EVTDESC_WCCLS_2151;You have waited so long, but have no answer from [mentor.GetBestName]. You don't know what happend, but it seems you need to look for another mentor.;;;;;;;;;;;;;x
+GetClassSkill_the_bow;the bow;;;;;;;;;;;;;x
```
> GetClassSkill_the_bow;the bow;;;;;;;;;;;;;x

Bow is not the only thing hunters can use, dwarves use rifles, for example.

mb replace it with "the hunt"?

arithon replied:
ranged weapon(ry)?

zumbak04 replied:
Ranged weapon sounds too long.
I would rather base it on your culture... or techs.

zumbak04 replied:
> Ranged weapon sounds too long.
> I would rather base it on your culture... or techs.

Done it in https://github.com/Warcraft-GoA-Development-Team/Warcraft-Guardians-of-Azeroth/pull/948/commits/1cba1705a2951e9e35eb58086958ed9623390762.

MemeWizard replied:
Tbh, I would prefer if it was the same for all hunters (marksmanship?) but ok

zumbak04 commented:
> > You no longer see **Ask to be Mentor** decisions by clicking evil characters.
> 
> What about vice versa?

@MemeWizard, I fixed it in https://github.com/Warcraft-GoA-Development-Team/Warcraft-Guardians-of-Azeroth/pull/1217. Could you check it?

zumbak04 commented:
> Suggestion:
> Add "master of dagger" for rogues

Done in https://github.com/Warcraft-GoA-Development-Team/Warcraft-Guardians-of-Azeroth/pull/948/commits/1cba1705a2951e9e35eb58086958ed9623390762.
